working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+
+and with an initializer list consisting of a single `super` invocation of either `super`, if *s* is named `S`, or `super.id`, if *s* is named `S.id`, with one argument for each parameter of *c* corresponding to a parameter of *s*, forwarding the value of the *c* parameter as an argument to the corresponding *s* parameter.
+
+The constructor *c* is `const` if *s* is `const` and *C* does not declare *any* instance variables.
+
+It is a *compile-time error* if there are no superclass constructors satisfying the requirements. (This is not a necessity, it's completely valid for a class to have no generative constructors, which is what currently happens if a class declares only factory constructors. We require the author to explicitly opt in to this behavior because *if* it is unintentional, it is a very hard-to-spot error.)
+
+It is a *compile-time error* to derive a mixin from a class unless its superclass is `Object`, it declares no generative constructors, and it declares no potentially non-nullable instance members without an initializer expression. (For NNBD-legacy code, it must declare no final fields without initializer expressions.)
+
+### Examples
+
+The following class declarations will be valid:
+
+```dart
+class Person {
+  final String firstName;
+  final String lastName;
+  final int age;
+}
+
+class Fireman extends Person {
+  final int yearsOfService;
+}
+```
+
+You can instantiate a `Fireman` as:
+
+```dart
+ Fireman(firstName: "John", lastName: "Doe", age: 37, yearsOfService: 12)
+```
+
+This is still verbose, because of the named arguments.
+
+You can extend classes with non-default constructors and forward implicitly to those as well:
+
+```dart
+class ColorBox extends Rectangle<int> {
+  final Color color;
+}
+... ColorBox(0, 10, 0, 10, color: Color.red) ...
+```
+
+This is not always as useful as it seems because it doesn't work when the superclass constructor has optional positional parameters. You cannot usefully extend `DateTime` the same way because the only constructor which doesn't have optional positional parameters is `DateTime.now`.
+
+It is still a very useful feature when you simply add methods to a class:
+
+```dart
+class MyRectangle extends Rectangle<int> {
+  MyRectangle affineTransform(int a, int b, int c, int d) => 
+      MyRectangle(a * left + b * top, c * left + d * top, 
+                  a * right + b * bottom, c * right + d * bottom);
+}
+```
+
+Here you implicitly "inherit" all constructors from the superclass without having to rewrite them.
```

Comment:
This implies that adding a constructor to a class will cause new constructors to spontaneously appear in all subclasses (that don't define their own constructors). If the superclass is one you don't control, you might not want your own class signing up for that extensibility.

In particular, this interacts really interestingly with package management. Consider a package foo:

### foo 1.0.0:

#### pubspec.yaml:

```yaml
name: foo
version: 1.0.0
```

#### lib/foo.dart:

```dart
class Foo {
  Foo.one();
}
```

Later, the maintainer adds another constructor:

### foo 1.1.0:

#### pubspec.yaml:

```yaml
name: foo
version: 1.1.0
```

#### lib/foo.dart:

```dart
class Foo {
  Foo.one();
  Foo.two();
}
```

This is a non-breaking change, so they just bump the minor version. Now consider a second package bar:

### bar 1.0.0:

#### pubspec.yaml:

```yaml
name: bar
version: 1.0.0
dependencies:
  - foo: '^1.0.0'
```

### lib/bar.dart:

```dart
class Bar extends Foo {}
```

In principle, the dependency constraint that bar places on foo is fine. bar should work with any version of foo compatible with foo 1.0.0.

But imagine some app that depends on bar 1.0.0. Can it call `Bar.second()`? Well, that depends on which version of foo got selected. If some other shared dependency on foo caused it to be pinned to 1.0.0, the answer is "no". Otherwise, the answer is "yes".

Do you think it's safer to limit this to the unnamed constructor? One option, if you like my suggestion for an explicit `= default` syntax is to only use the unnamed constructor implicitly and let users opt in to named ones using that syntax.

Reply:
That might be a good solution.
The *default* constructor keeps being the unnamed constructor, but you can declare *a* default constructor manually.

Reply:
This is not new, just different.

If my library v 1.0.0 declares:
```dart
  class C {}
```
and v1.1.0 adds:
```dart
  class C {
    static int foo() => 42;
  }
```
then someone depending on >=1.0.0 *may* get 1.1.0 and think that `C.foo` exists.
That's why we add `@Since("2.1")` to the platform libraries so that the analyzer can trigger on non-breaking features introduced after your min-dependency.

Versioning is not safe unless you lock yourself to *one* version, or always check your code against the *oldest* version you claim compatibility with.
It's not enough to use the oldest solution to the current constraints, because you might depend on some other package which depends on 1.1.0 of my package above. If they loosen their constraint to accept 1.0.0 later, then you can again end up being run on 1.0.0. So, only safe check is to analyze your code against the oldest supported version of each package that you use, even if they are not mutually compatible. (And that may not make sense at all, if packages are exporting types from each other, so maybe it's just impossible).

Reply:
> If my library v 1.0.0 declares:
> 
> ```dart
>   class C {}
> ```
> 
> and v1.1.0 adds:
> 
> ```dart
>   class C {
>     static int foo() => 42;
>   }
> ```
> 
> then someone depending on >=1.0.0 _may_ get 1.1.0 and think that `C.foo` exists.

This case is different. If you want to call `C.foo()`, you need to depend on the version of your package that includes it, 1.1.0.

My example involves *three* libraries. My app depends on bar, which transitively depends on foo. I have *no* direct dependency on foo at all. But with the current proposal, the set of constructors that *Bar* exposes depends on the version of *foo* that is selected with my app.